validphys2/src/validphys/pseudodata.py

```diff
@@ -198,38 +198,46 @@ def make_replica(
     full_mask=np.concatenate(check_positive_masks, axis=0)
     # The inner while True loop is for ensuring a positive definite
     # pseudodata replica
-    while True:
-        mult_shifts = []
-        # Prepare the per-dataset multiplicative shifts
-        for mult_uncorr_errors, mult_corr_errors in nonspecial_mult:
-        # convert to from percent to fraction
-            mult_shift = (
-                1 + mult_uncorr_errors * rng.normal(size=mult_uncorr_errors.shape) / 100
-            ).prod(axis=1)
-
-            mult_shift *= (
-                1 + mult_corr_errors * rng.normal(size=(1, mult_corr_errors.shape[1])) / 100
-            ).prod(axis=1)
-
-            mult_shifts.append(mult_shift)
-            
-        #If sep_mult is true then the multiplicative shifts were not included in the covmat
-        shifts = covmat_sqrt @ rng.normal(size=covmat.shape[1])
-        mult_part = 1.
-        if sep_mult:
-            special_mult = (
-                1 + special_mult_errors * rng.normal(size=(1, 
-                special_mult_errors.shape[1])) / 100
-                ).prod(axis=1)
-            mult_part = np.concatenate(mult_shifts, axis=0)*special_mult
-        #Shifting pseudodata
-        shifted_pseudodata = (all_pseudodata + shifts)*mult_part
-        #positivity control
-        if np.all(shifted_pseudodata[full_mask] >= 0):
-            break
 
+    mult_shifts = []
+    # Prepare the per-dataset multiplicative shifts
+    for mult_uncorr_errors, mult_corr_errors in nonspecial_mult:
+    # convert to from percent to fraction
+        mult_shift = (
+            1 + mult_uncorr_errors * rng.normal(size=mult_uncorr_errors.shape) / 100
+        ).prod(axis=1)
+
+        mult_shift *= (
+            1 + mult_corr_errors * rng.normal(size=(1, mult_corr_errors.shape[1])) / 100
+        ).prod(axis=1)
+
+        mult_shifts.append(mult_shift)
+            
+    #If sep_mult is true then the multiplicative shifts were not included in the covmat
+    shifts = covmat_sqrt @ rng.normal(size=covmat.shape[1])
+    mult_part = 1.
+    if sep_mult:
+        special_mult = (
+            1 + special_mult_errors * rng.normal(size=(1, 
+        special_mult_errors.shape[1])) / 100
+        ).prod(axis=1)
+        mult_part = np.concatenate(mult_shifts, axis=0)*special_mult
+    #regularize the negative shifts
+    reg_shifts = delta(shifts, all_pseudodata, full_mask)
+    #Shifting pseudodata
+    shifted_pseudodata = (all_pseudodata + reg_shifts)*mult_part
+          
     return shifted_pseudodata
 
+def delta(shifts, data, mask):
+    new_shifts = shifts.copy()
+    neg = shifts < 0.
+    negtofix = np.logical_and(neg, mask)
```

Comment:
You can’t count on the data being positive in the general case that includes closure tests. Also not clear to me it is the right scale to set (but maybe the uncertainty is?).

Reply:
Wait, how come that data are not positive?

In that case I would give up, and just regularize data with a `max(d, 0)`. I'm pretty sure we'll never get a negative central value from experimentalists, for a quantity supposed to be positive.

Reply:
> Also not clear to me it is the right scale to set (but maybe the uncertainty is?).

This actually I didn't understand, which scale are you talking of? However, the shape has been optimized for:

- mapping `0` shift in `0`
- mapping `-inf` to obtain `0` for the sampled datum (so the shift is `-datum`)
- have continuous first derivative (i.e. derivative equal to `1` for shift equal `0`)

Reply:
In principle we may want to not have to deal with experimental data but rather with theory predictions, which can logically (and practically) be negative. The most popular application of that is closure tests, but also the kind of thing we do in Cambridge these days (what happens if I add new physics to this pristine self consistent environment where the fake data has been obtained from the predictions?).

Reply:
Then I'm sorry, but you are asking for something more than the status quo: at the moment data are imposed positive by *retrying* strategy. This PR is not a generalization of that, but simply it's trying to be a more efficient (and possibly slightly less biasing) implementation of that.

If you want to allow for negative elements, then you should bypass the impose positivity (that was pre-existing to this PR), that is a generalization and it's very simple to add to both implementations: bypass it.
Still, it is not in the current scope, but not even farther because of this.

Reply:
The point is not to allow for negative pseudodata, but rather for negative "experimental" central values. Which works now albeit in a painstaking way.

Reply:
Ok, you're right, but then it might be even worse than the badness cited above: you can have a rate lower than 50% for a single point, so:
1. if it is an isolated feature, you can decide whatever you wish among `max(d, 0)` and discarding the point
2. if it is consistent over a set of points, then even the former was not a solution: you'd develop an incredibly high rejection rate

If you're in case 2., I would suggest dropping pseudodata positivity (and thus bypass the update above), because it's absurd to fit positive pseudodata for actual negative central values.
Pseudodata positivity is an assumption on the distribution, but it is not mandatory: if the thing you're fitting makes sense to be negative, then use appropriate distributions.

Reply:
I guess I am saying that I would consider a proposal to either clip negative values at zero or to do nothing and rely on other positivity constraining features to be potentially advantageous. But I think the proposed solution as is now breaks the aforementioned existing use case.

Reply:
> But I think the proposed solution as is now breaks the aforementioned existing use case.

I believe there is no existing use case available, because while theoretically possible, it requires an exponential time, so it's only useful when it is irrelevant.
The issue we attempted to face in this PR is: is it possible to have a technically light implementation for the current use case, namely positive central values? And the answer seems to be yes.

The scope was a brief technical PR, for a proper extension to negative central values (not supported by the status quo) let's open a new one. I guess we can make both behaviors available to the runcard (clipping and nothing).

Reply:
I'm in favor to expose to the runcard even current and nothing at pseudodata level, but not *retrying*, that is only bad.

```yaml
pseudodata_positivity: elu/null
central_values_positivity: clip/null
```
(keys are random, you know better than me where to fit it in the runcard)

Reply:
Of course choosing `elu` and `null` for central values makes no sense, but for the usual fit central values **are** positive, so you will see the difference only in the other cases.